File: beecrowd/c++/grafos/ex2528.cpp
#include <bits/stdc++.h>
using namespace std;

typedef vector<int> vi;
typedef vector<vi> vvi;

int distancia(int origem, int destino, int proibido, vvi grafo, int numVertices){
    vi visitados(numVertices, 0);
    visitados[origem] = 1;
    vi distancias(numVertices, -1);
    distancias[origem] = 0;
    queue<int> fila;
    fila.push(origem);

    while (!fila.empty()){
        int v_atual = fila.front();
        fila.pop();

        for (int vizinho : grafo[v_atual]){
            if(visitados[vizinho] == 0 and vizinho != proibido){
                visitados[vizinho] = 1;
                distancias[vizinho] = distancias[v_atual] + 1;
                fila.push(vizinho);
                if (vizinho == destino)
                    return distancias[vizinho];
            }
        }
    }
    return 0;
}

int main(){
    int n, m, a, b, c, r, e; 
    // n = número de vértices, m = número de arestas
    // a, b para ler cidades que são ligadas por uma aresta
    // c = cidade de origem, r = cidade de destino, e = cidade proibida
    while(scanf("%d %d", &n, &m) != EOF){
        vvi grafo(n);
        for(int i=0; i<m; i++){
            cin>>a>>b;
            grafo[a-1].emplace_back(b-1);
            grafo[b-1].emplace_back(a-1);
        }
        cin>>c>>r>>e;
        cout<<distancia(c-1, r-1, e-1, grafo, n)<<endl;
    }
}

File: beecrowd/python/grafos/ex2528.py
def rota(origem, destino, proibido, lista):
    visitados = [0] * len(lista)
    distancia = [-1] * len(lista)
    visitados[origem] = 1
    distancia[origem] = 0
    fila = [origem]
    
    while len(fila) != 0:
        v_atual = fila[0]
        del fila[0]
        
        for vizinho in lista[v_atual]:
            if visitados[vizinho] == 0 and vizinho != proibido:
                visitados[vizinho] = 1
                distancia[vizinho] = distancia[v_atual] + 1
                fila.append(vizinho)
                if vizinho == destino:
                    return distancia[vizinho]
    
while True:
    try:
        n, m = map(int, input().split())
        cidades = [[] for i in range(n)]
        for i in range(m):
            a, b = map(int, input().split())
            cidades[a-1].append(b-1)
            cidades[b-1].append(a-1)
        c, r, e = map(int, input().split())
        print(rota(c-1, r-1, e-1, cidades))
    except EOFError:
        break